fg21sim/extragalactic/pointsources/pointsources.py

```diff
@@ -0,0 +1,361 @@
+                        pix_tmp = hp.ang2pix(
+                            self.nside, x_r, y_r )
+                        hpmap[pix_tmp] += ps_lobe[i]
+
+            # Lobe2
+            lobe2_theta = lobe_maj * np.cos(lobe_ang + np.pi) * au.rad
+            lobe2_theta = theta + lobe2_theta.to(au.deg)
+            lobe2_phi = lobe_maj * np.sin(lobe_ang + np.pi) * au.rad
+            lobe2_phi = phi + lobe2_phi.to(au.deg)
+            # Focuses
+            lobe_c = np.sqrt(lobe_maj**2 - lobe_min**2)  # focus distance
+            F1_core_x = lobe_c
+            F1_core_y = 0
+            F2_core_x = -lobe_c
+            F2_core_y = 0
+            # draw
+            step = lobe_maj / 10
+            x = np.arange(-lobe_maj, lobe_maj + step, step)
+            y = np.arange(-lobe_min, lobe_min + step, step)
+            # Ellipse
+            for p in range(len(x)):
+                for q in range(len(y)):
+                    DistFocus1 = np.sqrt(
+                        (x[p] - F1_core_x)**2 + (y[q] - F1_core_y)**2)
+                    DistFocus2 = np.sqrt(
+                        (x[p] - F2_core_x)**2 + (y[q] - F2_core_y)**2)
+                    if (DistFocus1 + DistFocus2 <= 2 * lobe_maj):
+                        # rotation
+                        x_ang = (x[p] * au.rad).to(au.deg)
+                        y_ang = (y[q] * au.rad).to(au.deg)
+                        x_r = ((x_ang * np.cos(lobe_ang + np.pi) -
+                               y_ang * np.sin(lobe_ang + np.pi) +
+                               lobe2_theta).value / 180 * np.pi)
+                        y_r = ((x_ang * np.sin(lobe_ang + np.pi) +
+                               y_ang * np.cos(lobe_ang + np.pi) +
+                               lobe2_phi).value / 180 * np.pi)
+                        # Judge and Fill
+                        if x_r > np.pi:
+                            x_r -= np.pi
+                        elif x_r < 0:
+                            x_r += np.pi
+                        if y_r > 2 * np.pi:
+                            y_r -= 2 * np.pi
+                        elif y_r < 0:
+                            y_r += 2 * np.pi
+                        pix_tmp = hp.ang2pix(self.nside, x_r,y_r)
+                        hpmap[pix_tmp] += ps_lobe[i]
+            # Core
+            pix_tmp = hp.ang2pix(self.nside, ps_data['Theta (deg)'] / 180 *
+                                 np.pi, ps_data['Phi (deg)'] / 180 * np.pi)
+            ps_core = ps_flux_list[:, 0]
+            hpmap[pix_tmp] += ps_core
+
+        return hpmap
+
+    def draw_ps(self):
+        """
+        Read csv ps list file, and generate the healpix structure vector
+        with the respect frequency.
+        """
+        # Init
+        num_freq = len(self.freq)
+        npix = hp.nside2npix(self.nside)
+        hpmaps = np.zeros((npix,num_freq))
+
+        # load csv
+        for filepath in self.files:
+            ps_type, ps_data = self.read_csv(filepath)
+
```

Comment:
Move the *CSV data reading* part out of this function, ``draw_ps()`` only draws the PS images using already *loaded* data.